### DISCUSSION TITLE
Ideas to solve failed mesa import - "AttributeError: module 'numpy.typing' has no attribute 'NDArray'"

### DISCUSSION THREAD

**Discussion by AshleighPrice** [mesa/mesa#1826] [Q&A] [answered]
I'm trying to import Mesa (2.1.1) using Juptyer notebooks in ArcGIS Pro 3.13 (Python 3.9.18). on Import I an attribute error tracing to Line 54:   FloatCoordinate = Union[Tuple[float, float], npt.NDArray[float]]. 

Is it a possible data type error or could it be a dependency issue? (Code and screen shot below)

AttributeError                            Traceback (most recent call last)
In  [1]:
Line 1:     import mesa

File C:\Program Files\ArcGIS\Pro\bin\Python\envs\AgentBasedModel\lib\site-packages\mesa\__init__.py, in <module>:
Line 8:     import mesa.flat.visualization as visualization

File C:\Program Files\ArcGIS\Pro\bin\Python\envs\AgentBasedModel\lib\site-packages\mesa\flat\__init__.py, in <module>:
Line 3:     from mesa.space import *  # noqa

File C:\Program Files\ArcGIS\Pro\bin\Python\envs\AgentBasedModel\lib\site-packages\mesa\space.py, in <module>:
Line 54:    FloatCoordinate = Union[Tuple[float, float], npt.NDArray[float]]

AttributeError: module 'numpy.typing' has no attribute 'NDArray'
---------------------------------------------------------------------------

![image](https://github.com/projectmesa/mesa/assets/43358139/2e0c19ad-51c8-4441-8b94-20236d7d37f7)

**Comment by rht** [comment, marked as answer]
`numpy.typing.NDArray` is a recent addition to NumPy since 1.21: https://numpy.org/doc/stable/reference/typing.html#numpy.typing.NDArray. It's possible that the installed NumPy version bundled in that ArcGIS Pro is older than that. I haven't used that software: is it possible to update the NumPy version manually? If this is not possible, you can also manually edit line 54 of C:\Program Files\ArcGIS\Pro\bin\Python\envs\AgentBasedModel\lib\site-packages\mesa\space.py
to
```python
FloatCoordinate = Any
```

**Reply by AshleighPrice** [reply]
Answered - Thank You!
